workspaces/ui-v2/src/store/shapes/thunks.ts

```diff
@@ -0,0 +1,195 @@
+import { createAsyncThunk } from '@reduxjs/toolkit';
+import * as Sentry from '@sentry/react';
+
+import { IForkableSpectacle } from '@useoptic/spectacle';
+import { JsonLike } from '<src>/types';
+import {
+  convertSpectacleChangeToChangeType,
+  SpectacleChange,
+} from '../spectacleUtils';
+import { ShapeId, ReduxShape } from './types';
+
+const ShapeQuery = `
+query X($shapeId: ID! $sinceBatchCommitId: String) {
+  shapeChoices(shapeId: $shapeId) {
+    id
+    jsonType
+    asObject {
+      fields {
+        name
+        fieldId
+        shapeId
+        contributions
+        changes(sinceBatchCommitId: $sinceBatchCommitId) {
+          added
+          changed
+          removed
+        }
+      }
+    }
+    asArray {
+      shapeId
+    }
+  }
+}`;
+
+type SpectacleField = {
+  name: string;
+  fieldId: string;
+  shapeId: string;
+  contributions: Record<string, string>;
+  changes: SpectacleChange;
+};
+
+type SpectacleArray = {
+  shapeId: string;
+};
+
+type SpectacleShape =
+  | {
+      id: string;
+      jsonType: JsonLike.OBJECT;
+      asArray?: undefined;
+      asObject: {
+        fields: SpectacleField[];
+      };
+    }
+  | {
+      id: string;
+      jsonType: JsonLike.ARRAY;
+      asArray: SpectacleArray;
+      asObject?: undefined;
+    }
+  | {
+      id: string;
+      jsonType: Exclude<JsonLike, JsonLike.OBJECT | JsonLike.ARRAY>;
+
+      asArray?: undefined;
+      asObject?: undefined;
+    };
+
+type SpectacleShapeResult = {
+  shapeChoices: SpectacleShape[];
+};
+
+const fetchSpectacleShape = async (
+  spectacle: IForkableSpectacle,
+  shapeId: string,
+  sinceBatchCommitId?: string
+): Promise<SpectacleShapeResult> => {
+  const results = await spectacle.query<
+    SpectacleShapeResult,
+    { shapeId: string; sinceBatchCommitId?: string }
+  >({
+    query: ShapeQuery,
+    variables: { shapeId, sinceBatchCommitId },
+  });
+
+  if (results.errors || !results.data) {
+    console.error(results.errors);
+    throw new Error(JSON.stringify(results.errors));
+  }
+
+  return results.data;
+};
+
+const fetchShapesAndChildren = async (
```

Comment:
The biggest change here is that we are saving the data in a slightly different format - where we store each shape (but not recursively) - it's up to the selector to join the relevant information.

This is done because most likely want to access shape information by shapeId when we make changes, etc - storing the data in this format makes it easier to query for shapes, rather than traversing all the rootShapes.

I'm not sure if we need an index for fieldId or a fieldId -> shapeId index, but we can definitely add that in later if we find its necessary.